### ISSUE TITLE
Redis Sentinel for Distributed Cache Error

### ISSUE THREAD

**Issue by mohammed-essa** [StackExchange/StackExchange.Redis#1690]
Hi all,

I am trying to use the DistributedMemoryCache with StackExchangeRedisCache on .NET 5 and I am really struggling to get it to work. I cannot find samples of Redis Cache using the Redis Sentinel topology. If anyone can guide me with this, it will be really appreciated.

The error I am getting is: 

RedisConnectionException: UnableToConnect on 172.29.0.4:6379/Interactive, Initializing/NotStarted, last: NONE, origin: BeginConnectAsync, outstanding: 0, last-read: 5s ago, last-write: 5s ago, keep-alive: 60s, state: Connecting, mgr: 10 of 10 available, last-heartbeat: never, global: 0s ago, v: 2.2.4.27433

The redis sentinel is running in docker using the following:

 redis:
    image: 'bitnami/redis:latest'
    environment:
      - REDIS_REPLICATION_MODE=master
      - REDIS_PASSWORD=str0ng_passw0rd
    networks:
      - silicon-core-shared
    ports:
      - '6379'
  redis-slave:
    image: 'bitnami/redis:latest'
    environment:
      - REDIS_REPLICATION_MODE=slave
      - REDIS_MASTER_HOST=redis
      - REDIS_MASTER_PASSWORD=str0ng_passw0rd
      - REDIS_PASSWORD=str0ng_passw0rd
    ports:
      - '6379'
    depends_on:
      - redis
    networks:
      - silicon-core-shared
  redis-sentinel:
    image: 'bitnami/redis-sentinel:latest'
    environment:
      - REDIS_MASTER_PASSWORD=str0ng_passw0rd
    depends_on:
      - redis
      - redis-slave
    ports:
      - '26379-26381:26379'
    networks:
      - silicon-core-shared
volumes:
  redis_data:
    driver: local

The service collection config has the following:

 ConfigurationOptions redisOptions = new ConfigurationOptions();
            redisOptions.EndPoints.Add("localhost:26379"); 
            redisOptions.EndPoints.Add("localhost:26380"); 
            redisOptions.EndPoints.Add("localhost:26381");
            redisOptions.ServiceName = "mymaster";
            redisOptions.Ssl = false;
            redisOptions.TieBreaker = "";
            redisOptions.AllowAdmin = true;
            redisOptions.AbortOnConnectFail = false;
            redisOptions.Password = "str0ng_passw0rd";
            services.AddSingleton<IConnectionMultiplexer>(_ => ConnectionMultiplexer.Connect(redisOptions));
            services.AddDistributedMemoryCache();
            services.AddStackExchangeRedisCache(options =>
            {
                options.ConfigurationOptions = redisOptions;
            });

Thank you

**Comment by mgravell** [comment]
Note: the exact APIs you're using are in other libraries, so I'm a little
cautious in my advice, but: based on the exception, it looks like it simply
can't connect to the redis server. Is there line of sight, meaning:
firewalls, subnets, etc?

**Comment by mohammed-essa** [comment]
Thank you Marc for the assistance, I totally understand that this library is an external library. I am now able to connect to the redis sentinel from redis-commander in the same docker image. However I am now getting the error below when trying to read a value from the cache:

StackExchange.Redis.RedisConnectionException: 'No connection is active/available to service this operation: HMGET
UnableToConnect on redis-sentinel:6379/Interactive, Initializing/NotStarted, last: NONE, origin: BeginConnectAsync, outstanding: 0, last-read: 0s ago, last-write: 0s ago, keep-alive: 60s, state: Connecting, mgr: 10 of 10 available, last-heartbeat: never, global: 0s ago, v: 2.2.4.27433, mc: 1/1/0, mgr: 10 of 10 available

I am 100% sure that the sentinel is up and the slave nodes but I just cannot connect to it. The code I am using at Startup is:

services.AddSingleton<IConnectionMultiplexer>(_ => ConnectionMultiplexer.SentinelConnect(“sentinel-redis:26379,serviceName=mymaster”));
            services.AddDistributedMemoryCache();
            services.AddStackExchangeRedisCache(options =>
            {
                options.ConfigurationOptions = new ConfigurationOptions()
                {
                    AllowAdmin = true,
                    Password = appConfig.RedisOptions.Password,
                    ConnectTimeout = appConfig.RedisOptions.ConnectionTimeout,
                    AbortOnConnectFail = false,
                    EndPoints = { “mymaster”, 26379 }
                };
            });

Is there maybe something I am doing incorrectly? I just cannot figure this one out for some reason.

Thanks again for all the help on this.

**Comment by mgravell** [comment]
Asking obvious questions, but: is the address and port correct?

**Comment by mohammed-essa** [comment]
Yeah, double checked now in redis-commander and the project.

"RedisOptions": {
    "Host": "redis-sentinel",
    "Port": 26379,
    "Password": "str0ng_passw0rd",
    "Attempts": 5, //not used
    "Timeout": 5000,
    "ConnectionTimeout": 5000,
    "SentinelOptions": {
      "Master": "mymaster",
      "Nodes": 2
    }

  }

**Comment by mohammed-essa** [comment]
I tried the RedisClient as well now and also having a problem connecting. 

My appsettings are:

"Redis": {
    "Password": "str0ng_passw0rd",
    "AllowAdmin": true,
    "Ssl": false,
    "ConnectTimeout": 6000,
    "ConnectRetry": 2,
    "Hosts": [
      {
        "Host": "redis-sentinel",
        "Port": "26379"
      }
    ],
    "Database": 0
  }

**Comment by mgravell** [comment]
In the original error, you were connecting to :6379, not :23679 - you may
want to check that.

**Comment by mgravell** [comment]
Hmm; can you clarify: is the server also on the sentinel? Re-checking, I
see both 6379 and 23679

**Comment by mohammed-essa** [comment]
I agree Marc, that is the confusion for me as well, in the code I am using the below

ConfigurationOptions redisOptions = new ConfigurationOptions();
redisOptions.EndPoints.Add("localhost:26379");
redisOptions.EndPoints.Add("localhost:26380");
redisOptions.EndPoints.Add("localhost:26381");
redisOptions.ServiceName = "mymaster";
redisOptions.Ssl = false;
redisOptions.TieBreaker = "";
redisOptions.AllowAdmin = true;
redisOptions.AbortOnConnectFail = false;
redisOptions.Password = "str0ng_passw0rd";
services.AddSingleton(_ => ConnectionMultiplexer.Connect(redisOptions));
services.AddDistributedMemoryCache();
services.AddStackExchangeRedisCache(options =>
{
options.ConfigurationOptions = redisOptions;
});

but the error still says port 6379 which baffles me

**Comment by mgravell** [comment]
K. Does ConnectionMultiplexer.Connect with your setup work correctly? I
can't offer advice on anything outside this library, fundamentally.

**Comment by mohammed-essa** [comment]
no Marc, it does not. This is where I started. I tried to implement the below: 

ConfigurationOptions redisOptions = new ConfigurationOptions();
redisOptions.EndPoints.Add("localhost:26379");
redisOptions.ServiceName = "mymaster";
redisOptions.Ssl = false;
redisOptions.TieBreaker = "";
redisOptions.AllowAdmin = true;
redisOptions.AbortOnConnectFail = false;
redisOptions.Password = "str0ng_passw0rd";
services.AddSingleton(_ => ConnectionMultiplexer.Connect(redisOptions));
services.AddDistributedMemoryCache();
services.AddStackExchangeRedisCache(options =>
{
options.ConfigurationOptions = redisOptions;
});

I really tried a few options but no luck really. If at all possible that you have a code sample or snippet using the redis sentinel topology, I would really appreciate it.

Thank you
